core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,238 @@
+    mEvictLock = new ReentrantLock();
+    mLastLogTime = System.currentTimeMillis();
+  }
+
+  private void evictIfOverLimit() {
+    long numToEvict = mCache.size() - mSoftLimit;
+
+    if (numToEvict <= 0) {
+      return;
+    }
+
+    if (mEvictLock.tryLock()) {
+      try {
+        // This thread won the race as the evictor
+        while (numToEvict > 0) {
+
+          if (!mIterator.hasNext()) {
+            mIterator = mCache.entrySet().iterator();
+          }
+          Map.Entry<K, ValNode> candidateMapEntry = mIterator.next();
+          ValNode candidate = candidateMapEntry.getValue();
+          if (candidate.mIsAccessed) {
+            candidate.mIsAccessed = false;
+          } else {
+            if (candidate.mRefCount.compareAndSet(0, Integer.MIN_VALUE)) {
+              // the value object can be evicted, at the same time we make refCount minValue
+              mIterator.remove();
+              numToEvict--;
+            }
+          }
+        }
+      } finally {
+        mEvictLock.unlock();
+      }
+    }
+  }
+
+  /**
+   * get the value from the cache.
+   *
+   * @param key the key to look up the cache
+   * @return the value contained in the cache, if it is already in cache,
+   * otherwise generate an entry based on the loader
+   */
+
+  public LockResource get(final K key, LockResource.LockMode mode) {
+    ValNode valNode = getValNode(key, true);
+    ReentrantReadWriteLock lock = valNode.mValue;
+    switch (mode) {
+      case READ:
+        return new RefCountLockResource(lock.readLock(), valNode.mRefCount);
+      case WRITE:
+        return new RefCountLockResource(lock.writeLock(), valNode.mRefCount);
+      default:
+        throw new IllegalStateException("Unknown lock mode: " + mode);
+    }
+  }
+
+  public ReentrantReadWriteLock getRawLock(final K key) {
+    ValNode valNode = getValNode(key, false);
+    return valNode.get();
+  }
+
+  private ValNode getValNode(final K key, boolean refCount) {
+    Preconditions.checkNotNull(key, "key can not be null");
+    // oldCacheEntry keeps track of the last cache entry that was in the process of being removed.
+    // we do not need to do anything if we did not get a new CacheEntry
+    ValNode oldCacheEntry = null;
+    ValNode cacheEntry;
+    while (true) {
+      // repeat until we get a cacheEntry that is not in the process of being removed.
+      cacheEntry = mCache.compute(key, (k, v) -> {
+        if (v != null) {
+          v.getRefCounter().incrementAndGet();
+          v.mIsAccessed = true;
+          return v;
+        }
+        // new cache entry
+        if (mCache.size() >= mHardLimit) {
+          return null;
+        } else {
+          return new ValNode(mDefaultLoader.apply(k));
+        }
+      });
+
+      if (cacheEntry == null) {
+        // cache is at hard limit
+        try {
+          if (System.currentTimeMillis() - mLastLogTime > 60000) {
+            LOG.info("Cache at hard limit, size = " + mCache.size());
```

Comment:
I'd put this at warn level and log the hard and soft limits for reference

Reply:
done